acme/dns_challenge.go

```diff
@@ -262,9 +262,15 @@ func FindZoneByFqdn(fqdn string, nameservers []string) (string, error) {
 		if in.Rcode == dns.RcodeSuccess {
 			for _, ans := range in.Answer {
 				if soa, ok := ans.(*dns.SOA); ok {
-					zone := soa.Hdr.Name
-					fqdnToZone[fqdn] = zone
-					return zone, nil
+					if zone, err := publicsuffix.EffectiveTLDPlusOne(UnFqdn(fqdn)); err == nil {
```

Comment:
With the caveat that I am not a DNS protocol expect, this seems to be the best that can be done. First checking SOA records to ensure, this is a valid domain. Then if this is a known public suffix (known to the publicsuffix lib), then use that to determine zone. If this is unknown to public suffix lib (presumably a new TLD), then use the result from DNS server (but in that the result might be incorrect if CNAME is used).